cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/openai/OpenAI.scala

```diff
@@ -324,6 +324,62 @@ trait HasOpenAITextParams extends HasOpenAISharedParams {
   }
 }
 
+trait HasRAIContentFilter {
+  /**
+   * Determines if the content in the output row was filtered by content safety
+   * @param outputRow The output row from the API response
+   * @return true if content was filtered, false otherwise
+   */
+  def isContentFiltered(outputRow: Row): Boolean
+
+  /**
+   * Extracts the error/status reason from a filtered output row
+   * @param outputRow The output row from the API response
+   * @return The error reason (e.g., finish_reason or status)
+   */
+  def getFilterReason(outputRow: Row): String
+}
+
+trait HasChatOutput {
+  /**
+   * Extract the text content from the output column for this specific API type
+   */
+  def getOutputMessageString(outputColName: String): org.apache.spark.sql.Column
+}
+
+private[openai] object OpenAIMessageContentEncoder {
+
+  private def hasField(row: Row, fieldName: String): Boolean =
+    row.schema.fieldNames.contains(fieldName)
+
+  private def getOption[T](row: Row, fieldName: String): Option[T] =
+    if (hasField(row, fieldName)) Option(row.getAs[T](fieldName)) else None
+
+  private def castContentPartToString(part: Map[String, Any]): Map[String, Any] =
+    part.map { case (k, v) => k.toString -> v }
+
+  def rowsToMaps(messages: Seq[Row]): Seq[Map[String, Any]] = {
+    messages.map { row =>
+      val base = scala.collection.mutable.Map[String, Any]()
```

Comment:
avoid mutability in scala at all costs, avoid vars and mutable datastructyres, use maps to populate datastructures: aka

array.map(foo -> (foo, bar)).toMap